jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java

```diff
@@ -936,6 +970,70 @@ protected void runJob(Runnable job)
         job.run();
     }
 
+    /**
+     * <p>Attempts to shrink the current thread from the pool if {@link #getThreads()} is greater than {@link #getMinThreads()}.</p>
+     * @return true if the current thread was shrunk, false otherwise.
+     */
+    protected boolean shrink()
+    {
+        long idleTimeoutNanos = TimeUnit.MILLISECONDS.toNanos(getIdleTimeout());
+        while (true)
+        {
+            // No jobs available, should we shrink?
+            int threads = getThreads();
+            int minThreads = getMinThreads();
+
+            // This minThreads comparison isn't atomic, so there's a chance that multiple threads might
+            // enter this branch concurrently and shrink the pool below minThreads; this is okay
+            // as when that happens, the finally block in Runner.run() compensates b/c it calls
+            // ensureThreads() - so we can go below minThreads but that's a very transient state.
+            if (threads > minThreads)
+            {
+                // We have an idle timeout and excess threads, so check if we should shrink?
+                long now = NanoTime.now();
+                long shrinkPeriod = idleTimeoutNanos / getMaxShrinkCount();
+                if (LOG.isDebugEnabled())
+                    LOG.debug("Shrink check, {} > {} period={}ms {}", threads, minThreads, TimeUnit.NANOSECONDS.toMillis(shrinkPeriod), QueuedThreadPool.this);
+
+                long shrinkThreshold = _shrinkThreshold.get();
+                long threshold = shrinkThreshold;
+
+                // If the threshold is too far in the past,
+                // advance it to be one idle timeout before now plus a bit of shrink period.
+                if (NanoTime.elapsed(threshold, now) > idleTimeoutNanos)
+                    threshold = now - idleTimeoutNanos;
+
+                // advance the threshold by one shrink period
+                threshold += shrinkPeriod;
+
+                // Is the new threshold in the future?
+                if (NanoTime.isBefore(now, threshold))
+                {
+                    // Yes - we cannot shrink yet, so continue looking for jobs
+                    if (LOG.isDebugEnabled())
+                        LOG.debug("Shrink skipped, threshold={}ms in future {}", NanoTime.millisElapsed(now, threshold), QueuedThreadPool.this);
+                    return false;
+                }
+
+                // We can shrink if we can update the atomic shrink threshold?
+                if (_shrinkThreshold.compareAndSet(shrinkThreshold, threshold))
+                {
+                    // Yes - we have shrunk, so exit.
+                    if (LOG.isDebugEnabled())
+                        LOG.debug("SHRUNK, threshold={}ms in past {}", NanoTime.millisElapsed(threshold, now), QueuedThreadPool.this);
```

Comment:
Remove the upper case.

```suggestion
                        LOG.debug("Shrunk, threshold={}ms in past {}", NanoTime.millisElapsed(threshold, now), QueuedThreadPool.this);
```

However, if we agree on "evict", these log statements should be changed.